src/D2L.CodeStyle.Analyzers/Async/Generator/FileCollector.cs

```diff
@@ -0,0 +1,273 @@
+﻿using System.Collections.Immutable;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace D2L.CodeStyle.Analyzers.Async.Generator;
+
+internal partial class SyncGenerator {
+	/// <summary>
+	/// Creates a source file containing generated methods and the necessary
+	/// namespaces, using directives, and type declarations for them to live in
+	/// (but only the necessary ones).
+	/// </summary>
+	internal sealed class FileCollector {
+		private static readonly SyntaxToken NoneToken
+			= SyntaxFactory.Token( SyntaxKind.None );
+
+		private static readonly SyntaxTriviaList EmptyTrivia
+			= SyntaxFactory.TriviaList();
+
+		private readonly CompilationUnitSyntax m_root;
+
+		// This is a mutable dictionary because we will remove from it as we go
+		// and check that it is empty at the end, just to be safe.
+		private readonly Dictionary<SyntaxNode, IEnumerable<string>> m_methods;
+
+		// When we recurse into a namespace or type, we're not sure if the
+		// preamble ("namespace foo {", "\t\tpartial class Bar<T> {", etc.) is
+		// necessary, because maybe there are no generated methods.
+		// Using a LinkedList<string> as a double-ended queue.
+		private readonly LinkedList<string> m_preambles = new();
+
+		private readonly StringBuilder m_out = new();
+
+		private FileCollector(
+			CompilationUnitSyntax root,
+			Dictionary<SyntaxNode, IEnumerable<string>> methods
+		) {
+			m_root = root;
+			m_methods = methods;
+		}
+
+		public static FileCollector Create(
+			CompilationUnitSyntax root,
+			ImmutableArray<(SyntaxNode Parent, string Source)> methods
+		) {
+			var groupedMethods = methods
+				.GroupBy( static m => m.Parent )
+				.ToDictionary(
+					static g => g.Key,
+					static g => g.Select( static m => m.Source )
+				);
+
+			return new FileCollector( root, groupedMethods );
+		}
+
+		public string CollectSource() {
+			// File-scoped usings:
+			m_out.Append( m_root.Usings.ToFullString() );
+
+			WriteChildren( m_root );
+
+			if( m_preambles.Count != 0 ) {
+				throw new Bug( "left over preambles" );
+			}

```

Comment:
These caught some bugs while I was writing it, so I left it in place. It's cheap and it makes it easier to debug tests too.

Reply:
These `FileCollector.Bug` exceptions will get caught and turned into diagnostics to not anger VS.